### PR TITLE
[DONOTMERGE] suggestion: new batch/requestheader protos

### PR DESCRIPTION
**Pull request by tbg**
shows suggestion for #2143 for refactoring the incoming request path (outgoing is easier).

### CODE REVIEW

**Comment by tschottdorf** [comment]
Getting rid of the non-txn timestamps seems to work, except that there's a caveat when applying a batch from Raft: If keys are accessed multiple times (i.e. write twice), this must not happen at a per-batch timestamp (or the second write goes bust).

A solution could be the following:
```go
baseTS := hlc.Now() // in `AddCmd`
allocTS := baseTS.Add(0, len(batch.Requests)) // one tick for each request
// now use `allocTS` for `{end,begin}Cmd` (so that the timestamp cache reports
// something higher than each command's timestamp), but actually execute request `i`
// with timestamp `baseTS.Add(i)`.
```

Of course, the actual implementation should be a little less ad-hoc, but this could be a workable idea.

**Comment by tschottdorf** [comment]
Alternatively, we could enforce that `BatchChunk` never overlaps itself when it's not in a `Txn`. But that seems odd.

File: proto/refactored.proto [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+//            applyRaftCommandInBatch checks that the lease covers the timestamp
+//
+// All of those would work with one timestamp per Batch. When executing a Batch
+// outside of a Txn, we'll need to do something though, otherwise a Batch whose
+// requests have overlapping keys will likely fail, for example writing a key
+// twice. That must not happen at the same timestamp. We need to "abuse" the
+// logical component for this.
+
+// KeyHeader replaces RequestHeader.
+message KeyHeader {
```
Comment (vivekmenezes) [inline proto/refactored.proto]: Should we call this Range?

**Comment by tamird** [comment]
I'm principally on board with this.

File: proto/refactored.proto (hunk 1 above)
Comment (vivekmenezes) [inline proto/refactored.proto]: We shouldn't call it a range to mistake it with a cockroach range.

**Comment by mrtracy** [comment]
LGTM

**Comment by tschottdorf** [comment]
update to my last comment: There could be some unforeseen consequences with the above `Timestamp` suggestion.

Imagine a `Txn` writing 10 intents. Now a `Scan` comes along. It gets a `WriteIntentError` with those 10 intents and initiates a batch worth of `PushTxn`. Call the timestamp in that Batch `T1`.

Now `T1` executes with the above, so we basically push the transaction 10 times with increasing timestamps. That's not going to end well, since each Push will randomly run against the Txn's priority and that will increase to (almost) our own, so that we're likely to fail somewhere in the course of those ten Pushes.
This means a new attempt, and of course, same outcome.

Of course, we don't have to `Push` 10 times. Once is enough, with the maximum timestamp. But while we can fix this one case, better to think twice about other places where something like this might happen.

**Comment by bdarnell** [comment]
Non-transactional batches will be rare with everything going through SQL, so we should just do whatever is easier here. I think it's fine to disallow non-transactional batches that touch the same key more than once. The order of operations in a batch is already unspecified, so it doesn't really make sense to try this.

**Comment by spencerkimball** [comment]
:+1: to @bdarnell's comment. Let's simplify by disallowing any non-required use cases wherever possible.

**Comment by tschottdorf** [comment]
going to close this since it's not really a PR, but will work towards this in #2143.

File: proto/refactored.proto (hunk 1 above)
Comment (tamird) [inline proto/refactored.proto]: fwiw this type exists now (but is not a proto) and is called `Span`

Comment (tschottdorf) [inline proto/refactored.proto]: yep, ran into that and will use that. Thanks for the heads up though.

**Pull request closed without merging**